File: app/src/main/java/org/epstudios/epmobile/DateCalculator.java
package org.epstudios.epmobile;

import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.CheckBox;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.TextView;
import android.widget.Toast;

import java.text.DateFormat;
import java.util.Calendar;
import java.util.GregorianCalendar;

public class DateCalculator extends EpActivity implements OnClickListener {
    private DatePicker indexDatePicker;
    private RadioGroup dayRadioGroup;
    private EditText numberOfDaysEditText;
    private TextView calculatedDateTextView;
    private CheckBox reverseTimeCheckBox;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.daycalculator);
        setupInsets(R.id.my_root_view);
        initToolbar();

        View calculateDateButton = findViewById(R.id.calculate_button);
        calculateDateButton.setOnClickListener(this);
        View clearButton = findViewById(R.id.clear_button);
        clearButton.setOnClickListener(this);

        indexDatePicker = findViewById(R.id.indexDatePicker);
        dayRadioGroup = findViewById(R.id.dayRadioGroup);
        numberOfDaysEditText = findViewById(R.id.numberOfDaysEditText);
        reverseTimeCheckBox = findViewById(R.id.reverseTimeCheckBox);
        calculatedDateTextView = findViewById(R.id.calculated_date);
        numberOfDaysEditText.setText(R.string.dc_default_number_of_days);

        dayRadioGroup
                .setOnCheckedChangeListener((group, checkedId) -> {
                    RadioButton checkedRadioButton = group
                            .findViewById(checkedId);
                    int index = group.indexOfChild(checkedRadioButton);
                    int number = 0;
                    switch (index) {
                        case 0:
                            number = 90;
                            break;
                        case 1:
                            number = 40;
                            break;
                        case 2:
                            number = 30;
                            break;
                        // else still = 0;
                    }
                    if (number != 0)
                        numberOfDaysEditText.setText(String.valueOf(number));
                });

    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            Intent parentActivityIntent = new Intent(this, CalculatorList.class);
            parentActivityIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP
                    | Intent.FLAG_ACTIVITY_NEW_TASK);
            startActivity(parentActivityIntent);
            finish();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onClick(View v) {
        final int id = v.getId();
        if (id == R.id.calculate_button) {
            calculateDays();
        } else if (id == R.id.clear_button) {
            clearEntries();
        }
    }

    private void calculateDays() {
        CharSequence numberOfDays = numberOfDaysEditText.getText();
        try {
            int number = Integer.parseInt(numberOfDays.toString());
            if (reverseTimeCheckBox.isChecked())
                number = -number;
            Calendar cal = new GregorianCalendar(indexDatePicker.getYear(),
                    indexDatePicker.getMonth(), indexDatePicker.getDayOfMonth());
            cal.add(Calendar.DATE, number);
            // DateFormat =
            // SimpleDateFormat dateFormat = new SimpleDateFormat("MM/dd/yyyy");
            String message = DateFormat.getDateInstance(DateFormat.MEDIUM).format(cal.getTime());
            calculatedDateTextView.setText(message);
            Toast.makeText(this, message, Toast.LENGTH_LONG).show();
        } catch (NumberFormatException e) {
            calculatedDateTextView.setText(getString(R.string.invalid_warning));
            calculatedDateTextView.setTextColor(Color.RED);
        }

    }

    private void clearEntries() {
        numberOfDaysEditText.setText(null);
        calculatedDateTextView.setText(getString(R.string.date_result_label));
        dayRadioGroup.check(R.id.ninetyRadio);
        numberOfDaysEditText.setText(getString(R.string.dc_default_number_of_days));
        calculatedDateTextView.setTextAppearance(this,
                android.R.style.TextAppearance_Large);

    }

    @Override
    protected boolean hideInstructionsMenuItem() {
        return false;
    }

    @Override
    protected void showActivityInstructions() {
        showAlertDialog(R.string.date_calculator_title,
                R.string.date_calculator_instructions);
    }
}

File: app/src/main/java/org/epstudios/epmobile/features/calculators/ui/DateCalculator.kt
package org.epstudios.epmobile.features.calculators.ui

import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import android.widget.Toast
import com.google.android.material.datepicker.MaterialDatePicker
import org.epstudios.epmobile.R
import org.epstudios.epmobile.core.ui.base.EpActivity
import org.epstudios.epmobile.databinding.DatecalculatorBinding
import java.text.DateFormat
import java.util.Calendar
import java.util.GregorianCalendar
import java.util.TimeZone

class DateCalculator : EpActivity() {
    private lateinit var binding: DatecalculatorBinding
    private var selectedDate: Calendar = Calendar.getInstance()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = DatecalculatorBinding.inflate(layoutInflater)
        setContentView(binding.root)
        setupInsets(binding.root)
        initToolbar()

        binding.calculateButtonsLayout.calculateButton.setOnClickListener { calculateDays() }
        binding.calculateButtonsLayout.clearButton.setOnClickListener { clearEntries() }
        binding.indexDateButton.setOnClickListener { showDatePicker() }

        binding.numberOfDaysEditText.setText(R.string.dc_default_number_of_days)
        updateDateButtonText()

        binding.dayRadioGroup.setOnCheckedChangeListener {
            _, checkedId ->
            val days = when (checkedId) {
                R.id.ninetyRadio -> "90"
                R.id.fortyRadio -> "40"
                R.id.thirtyRadio -> "30"
                else -> null
            }
            days?.let {
                binding.numberOfDaysEditText.setText(it)
            }
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            android.R.id.home -> {
                val parentActivityIntent = Intent(this, CalculatorList::class.java)
                parentActivityIntent.addFlags(
                    Intent.FLAG_ACTIVITY_CLEAR_TOP
                            or Intent.FLAG_ACTIVITY_NEW_TASK
                )
                startActivity(parentActivityIntent)
                finish()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun calculateDays() {
        val numberOfDays = binding.numberOfDaysEditText.text.toString()
        try {
            var number = numberOfDays.toInt()
            if (binding.reverseTimeCheckBox.isChecked) number = -number
            val cal: Calendar = GregorianCalendar(
                selectedDate.get(Calendar.YEAR),
                selectedDate.get(Calendar.MONTH),
                selectedDate.get(Calendar.DAY_OF_MONTH)
            )
            cal.add(Calendar.DATE, number)
            val message = DateFormat.getDateInstance(DateFormat.MEDIUM).format(cal.time)
            binding.calculatedDate.text = message
            Toast.makeText(this, message, Toast.LENGTH_LONG).show()
        } catch (_: NumberFormatException) {
            binding.calculatedDate.text = getString(R.string.invalid_warning)
            binding.calculatedDate.setTextAppearance(R.style.TextAppearance_Calculator_Error)
        }
    }

    private fun clearEntries() {
        binding.numberOfDaysEditText.text = null
        binding.calculatedDate.text = getString(R.string.date_result_label)
        binding.dayRadioGroup.check(R.id.ninetyRadio)
        binding.numberOfDaysEditText.setText(getString(R.string.dc_default_number_of_days))
        binding.calculatedDate.setTextAppearance(R.style.TextAppearance_Calculator_Result)
        selectedDate = Calendar.getInstance()
        updateDateButtonText()
    }

    override fun hideInstructionsMenuItem(): Boolean {
        return false
    }

    override fun showActivityInstructions() {
        showAlertDialog(
            R.string.date_calculator_title,
            R.string.date_calculator_instructions
        )
    }

    private fun showDatePicker() {
        val picker = MaterialDatePicker.Builder.datePicker()
            .setTitleText(R.string.select_date_title)
            .setSelection(utcMillisFromCalendar(selectedDate))
            .build()
        picker.addOnPositiveButtonClickListener { selection ->
            selection?.let {
                selectedDate = calendarFromUtc(it)
                updateDateButtonText()
            }
        }
        picker.show(supportFragmentManager, "date-picker")
    }

    private fun updateDateButtonText() {
        val dateText = DateFormat.getDateInstance(DateFormat.MEDIUM).format(selectedDate.time)
        binding.indexDateButton.text = dateText
    }

    private fun calendarFromUtc(utcMillis: Long): Calendar {
        val utcCalendar = Calendar.getInstance(TimeZone.getTimeZone("UTC"))
        utcCalendar.timeInMillis = utcMillis
        val localCalendar = Calendar.getInstance()
        localCalendar.set(
            utcCalendar.get(Calendar.YEAR),
            utcCalendar.get(Calendar.MONTH),
            utcCalendar.get(Calendar.DAY_OF_MONTH)
        )
        return localCalendar
    }

    private fun utcMillisFromCalendar(calendar: Calendar): Long {
        val utcCalendar = Calendar.getInstance(TimeZone.getTimeZone("UTC"))
        utcCalendar.set(
            calendar.get(Calendar.YEAR),
            calendar.get(Calendar.MONTH),
            calendar.get(Calendar.DAY_OF_MONTH)
        )
        return utcCalendar.timeInMillis
    }
}
